Source/WebKit/UIProcess/API/gtk/WebKitWebViewBase.cpp

```diff
@@ -968,6 +975,43 @@ static void webkitWebViewBaseUnmap(GtkWidget* widget)
     webkitWebViewBaseScheduleUpdateActivityState(webViewBase, ActivityState::IsVisible);
 }
 
+static bool shouldForwardKeyEvent(WebKitWebViewBase* webViewBase, GdkEvent* event)
+{
+#if USE(GTK4)
+    return event && webViewBase->priv->keyEventsToPropagate.removeFirst(event);
+#else
+    return std::exchange(webViewBase->priv->shouldForwardNextKeyEvent, false);
+#endif
+}
+
+static bool shouldForwardWheelEvent(WebKitWebViewBase* webViewBase, GdkEvent* event)
+{
+#if USE(GTK4)
+    if (!event || webViewBase->priv->wheelEventsToPropagate.isEmpty())
+        return false;
+    if (gdk_event_get_event_type(event) != GDK_TOUCHPAD_HOLD) {
+        // Some events that were meant for delayed propagation may have been coalesced/compressed by GDK into the current event:
+        // https://docs.gtk.org/gtk4/migrating-3to4.html#adapt-to-gdkevent-api-changes
+        // There's no way to both propagate and stop the current event, so just clear out these obsolete events from the to-be-propagated vector.
+        unsigned length = 0;
+        GUniquePtr<GdkTimeCoord> history(gdk_event_get_history(event, &length));
+        if (history.get()) {
+            // The last entry in the history is the current event time, so ignore that.
+            for (unsigned i = 0; i < length - 1; i++) {
+                auto oldTime = history.get()[i].time;
+                webViewBase->priv->wheelEventsToPropagate.removeFirstMatching([&oldTime] (GRefPtr<GdkEvent>& current) {
+                    auto isMatch = gdk_event_get_time(current.get()) == oldTime;
+                    return isMatch;
+                });
+            }
```

Comment:
I don't understand what's happening here. The comment makes it look like you're trying to remove the last history entry for each event, but in fact your code is looking at all pending events and removing them if the event matches any time in the history. Is that intended? Why?

Reply:
Changed the code a bit to try to make it more clear the comment just pertains to the loop end condition adjustment.

The comment describing what this whole block is trying to do is a few lines above:
```
// Some events that were meant for delayed propagation may have been coalesced/compressed by GDK into the current event:
// https://docs.gtk.org/gtk4/migrating-3to4.html#adapt-to-gdkevent-api-changes
// There's no way to both propagate and stop the current event, so just clear out these obsolete events from the to-be-propagated vector.
```

So sometimes PageClientImpl tries to "propagate" an event after-the-fact, but when it adds the event to the GDK event queue, GDK ends up coalescing the event with some other active scroll event, meaning we'll never see the event we want to propagate in handleScroll(). So this code detects if an event that was meant for propagation was coalesced with the current event, and if so cleans up wheelEventsToPropagate so that we don't leak memory.